### PR TITLE
Apply resources in topological sort order

### PR DESCRIPTION
**Pull request by gibbleyg**

### CODE REVIEW

File: mpdev/cmd/applycmd.go [hunk 1]
```diff
@@ -56,29 +56,18 @@ func (c *command) RunE(_ *cobra.Command, _ []string) error {
...
 
-	// TODO(https://github.com/GoogleCloudPlatform/marketplace-tools/issues/8): Setup execution order for apply command
-	// Some resources cannot be applied prior to others
-	for _, resource := range resources {
-		err := resource.Apply()
-		if err != nil {
-			return err
-		}
-	}
+	err := registry.Apply()
```
Comment (huyhg) [inline mpdev/cmd/applycmd.go:68]: You can add `fixes #8` in the description of this PR to automatically close the issue.

File: mpdev/internal/apply/registry.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+// Apply invokes `Apply` on all resources in the registry.
+func (r *registry) Apply() error {
+	resources, err := r.topologicalSort()
+	if err != nil {
+		return err
+	}
+
+	for _, resource := range resources {
+		err = resource.Apply(r)
```
Comment (huyhg) [inline mpdev/internal/apply/registry.go:55]: maybe add a TODO/comment to consider parallel applies in the future.

File: mpdev/internal/apply/registry.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+// topologicalSort returns a list of resources such that each
+// resource is after its dependencies in the list.
+func (r *registry) topologicalSort() ([]Resource, error) {
```
Comment (huyhg) [inline mpdev/internal/apply/registry.go:66]: Can we add a unit test for this?

File: mpdev/internal/apply/registry.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		to := dag.Node(refToID[ref])
+		for _, depRef := range resource.GetDependencies() {
+			from := dag.Node(refToID[depRef])
+			e := dag.NewEdge(from, to)
+			dag.SetEdge(e)
+		}
+	}
+
+	// Execute topological sort
+	var resources []Resource
```
Comment (huyhg) [inline mpdev/internal/apply/registry.go:90]: nit: Move this down to L96 and provide it an initial capacity.

```go
resources := make([]Resource, 0, len(nodes))
```

File: mpdev/internal/apply/resource.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+// specification. The combination of Group, Kind, Name MUST be unique for all
+// applied resources.
+type Reference struct {
+	Group string
+	Kind  string
+	Name  string
+}
+
+// ReferenceMap is a mapping between KRM references and the resource object.
+type ReferenceMap map[Reference]Resource
```
Comment (huyhg) [inline mpdev/internal/apply/resource.go:38]: rm?

File: mpdev/internal/apply/registry.go (hunk 3 above)
Comment (gibbleyg) [reply mpdev/internal/apply/registry.go:66]: Yeah, will add in separate pr

File: mpdev/internal/apply/resource.go (hunk 5 above)
Comment (gibbleyg) [reply mpdev/internal/apply/resource.go:38]: Changing to `refMap`, rm seems to short to me

**Comment by gibbleyg** [comment]
@googlebot rescan

**Pull request merged**